Makefile

```
@@ -27,32 +27,46 @@ escape_double_quote = $(subst $\",$\\",$(1))
 # Utility function to require a variable to be defined
 require_defined = $(if $(value $1),$(value $1),$(error "$1 needs to be defined"))
 
+# What implementation to build
+VARIANT ?= trivial
 # Where to put build files
 OUT ?= out
 # Where to put typescript files
 TS_OUT ?= typescript
 
 .PHONY: all clean test typescript
 
-all: \
-	$(OUT)/baseDAO.tz
+all: $(OUT)/trivialDAO.tz $(OUT)/registryDAO.tz $(OUT)/treasuryDAO.tz $(OUT)/testDAO.tz
+
+trivialDAO: VARIANT = trivial
+trivialDAO: baseDAO
+
+registryDAO: VARIANT = registry
+registryDAO: baseDAO
+
+treasuryDAO: VARIANT = treasury
+treasuryDAO: baseDAO
+
+testDAO: VARIANT = test
+testDAO: baseDAO
 
 # Compile LIGO contract into its michelson representation.
-$(OUT)/baseDAO.tz: src/**
+$(OUT)/$(VARIANT)DAO.tz: src/**
+	cp src/$(VARIANT)DAO.mligo src/implementation.mligo
```

```suggestion
$(OUT)/%DAO.tz: src/**
	cp src/$*DAO.mligo src/implementation.mligo
```
this can be changed to use a wildcard (`%`), which will match any non-empty string.

Note that patterns are expanded first, so unless `OUT` is changed from the default value this will only match if the destination is in `out`.
E.g. these would work:
```sh
make out/trivialDAO.tz
make different_dir/trivialDAO.tz OUT=different_dir
```
but this would not;
```sh
make different_dir/trivialDAO.tz
```

Here I also used the automatic `$*` variable that expands to what `%` matched on, see [the docs](https://www.gnu.org/software/make/manual/html_node/Automatic-Variables.html).